src/main/kotlin/com/openlattice/data/ids/jobs/FixDuplicateIdAssignmentJob.kt

```diff
@@ -0,0 +1,307 @@
+        ps.setString(3, entityKey.entityId)
+    }
+
+    private fun recordCollision(id: UUID, entityKeys: List<EntityKey>) {
+        hds.connection.use { connection ->
+            connection.prepareStatement(INSERT_COLLISION).use { ps ->
+                ps.setObject(1, id)
+                entityKeys.forEach {
+                    ps.setObject(2, it.entitySetId)
+                    ps.setObject(3, it.entityId)
+                    ps.addBatch()
+                }
+                ps.executeBatch()
+            }
+        }
+    }
+
+    override fun initialize() {
+        state.idCount = getIdCount()
+    }
+
+    override fun updateProgress() {
+        progress = ((100 * state.idsProceessed) / state.idCount).toByte()
+    }
+
+    private fun getIdCount(): Long {
+        return hds.connection.use { connection ->
+            connection.createStatement().use { statement ->
+                val rs = statement.executeQuery(COUNT_SYNC_IDS)
+                if (rs.next()) {
+                    ResultSetAdapters.count(rs)
+                } else {
+                    throw IllegalStateException("Result set for counting ids cannot empty.")
+                }
+            }
+        }
+    }
+
+    @JsonIgnore
+    override fun setHikariDataSource(hds: HikariDataSource) {
+        this.hds = hds
+    }
+
+    @JsonIgnore
+    override fun setIdGenerationService(idService: HazelcastIdGenerationService) {
+        this.idService = idService
+    }
+}
+
+private const val DUPLICATES_FIELD = "duplicates"
+private val COUNT_SYNC_IDS = """
+SELECT count(*) FROM ${SYNC_IDS.name}
+""".trimIndent()
+
+/**
+ * Returns the duplicates for a batch of ids.
+ * 1. ids - array of uuids
+ */
+private val SELECT_NEXT_DUPLICATES = """
+SELECT  ${ID.name}, 
+        jsonb_agg(jsonb_build_object(
+            '${SerializationConstants.ENTITY_SET_ID}', ${ENTITY_SET_ID.name},
+            '${ENTITY_SET_ID.name}${SerializationConstants.ENTITY_SET_ID}',${ENTITY_ID.name})
+        ) as $DUPLICATES_FIELD 
+FROM ${SYNC_IDS.name} where ${ID.name} = ANY(?) 
+GROUP BY ${ID.name}
+HAVING count(${ENTITY_ID.name}) > 1
+""".trimIndent()
+
+/**
+ * Returns the next batch of ids of size [BATCH_SIZE]. Proof: If a partition contains an element
+ * smaller than the max id returned from the query then it would have been returned in the query instead of the max element.
+ *
+ * 1. id - uuid
+ */
+private val SELECT_NEXT_IDS = """
+SELECT DISTINCT ${ID.name} FROM ${SYNC_IDS.name} WHERE ${ID.name} > ? LIMIT $BATCH_SIZE)
+""".trimIndent()
+
+/**
+ * 1. entity set id - uuid
+ * 2. id - uuid
+ */
+private val MIGRATE_IDS = """
+WITH for_migration as (DELETE FROM ${IDS.name} WHERE ${ENTITY_SET_ID.name} = ? AND ${ID.name} = ? RETURNING *)
+    INSERT INTO ${IDS.name} 
+        SELECT $REPARTITION_IDS_COLUMNS 
+        FROM for_migration 
+        INNER JOIN (select ${ID.name} as ${ENTITY_SET_ID.name}, ${PARTITIONS.name} FROM ${ENTITY_SETS.name} ) as es 
```

Comment:
is this joining on partitions? how does postgres compare those?

Reply:
The USING clause will join on the entity_set_id field. Not that it matters since this we don't need this query anymore.